contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,959 @@
+    ///      strategy contract. For that reason any liquidity withdrawals also collect the tokens.
+    function _updateUnderlyingAssets() internal {
+        if (tokenId == 0) {
+            underlyingAssets = 0;
+        } else {
+            uint128 _liquidity = _getLiquidity();
+
+            /**
+             * Our net value represent the smallest amount of tokens we are able to extract from the position
+             * given our liquidity.
+             *
+             * The least amount of tokens extraditable from the position is where the active trading price is
+             * at the ticker 0 meaning the pool is offering 1:1 trades between WETH & OETHb. At that moment the pool
+             * consists completely of OETHb and no WETH.
+             *
+             * The more swaps from WETH -> OETHb happen on the pool the more the price starts to move towards the -1
+             * ticker making OETHb (priced in WETH) more expensive.
+             */
+            (uint256 _wethAmount, uint256 _oethbAmount) = helper
+                .getAmountsForLiquidity(
+                    sqrtRatioX96TickClosestToParity, // sqrtRatioX96
+                    sqrtRatioX96Tick0, // sqrtRatioAX96
+                    sqrtRatioX96Tick1, // sqrtRatioBX96
+                    _liquidity
+                );
+
+            require(_wethAmount == 0, "Non zero wethAmount");
+            underlyingAssets = _oethbAmount;
+        }
+    }
+
+    /**
+     * @notice Deposit WETH to the strategy contract. This function does not add liquidity to the
+     *         underlying Aerodrome pool.
+     */
+    function depositAll() external override onlyVault nonReentrant {
+        uint256 _wethBalance = IERC20(WETH).balanceOf(address(this));
+        if (_wethBalance > 0) {
+            _deposit(WETH, _wethBalance);
+        }
+    }
+
+    /**
+     * @notice Withdraw an `amount` of assets from the platform and
+     *         send to the `_recipient`.
+     * @param _recipient  Address to which the asset should be sent
+     * @param _asset      WETH address
+     * @param _amount     Amount of WETH to withdraw
+     */
+    function withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _amount
+    ) external override onlyVault nonReentrant {
+        require(_asset == WETH, "Unsupported asset");
+        require(_recipient == vaultAddress, "Only withdraw to vault allowed");
+
+        uint256 _wethBalance = IERC20(WETH).balanceOf(address(this));
+        if (_wethBalance < _amount) {
+            uint256 _additionalWethRequired = _amount - _wethBalance;
+            (uint256 _wethInThePool, ) = getPositionPrincipal();
+
+            if (_wethInThePool < _additionalWethRequired) {
+                revert NotEnoughWethLiquidity(
+                    _wethInThePool,
+                    _additionalWethRequired
+                );
+            }
+
+            uint256 shareOfWethToRemove = _additionalWethRequired.divPrecisely(
```

Comment:
This could break if we try to withdraw nearly all liquidity in the pool?

Reply:
good point thanks. Fixed here: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/87c163c2f9d7b4cc4bbeef18ba5b6564a84eac74